universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java

```diff
@@ -189,23 +189,38 @@ public ProgramDefinition moveBlock(long blockId, Direction direction)
         blockSlice.startsBefore(otherBlockSlice) ? blockSlice : otherBlockSlice;
     BlockSlice latterSlice =
         blockSlice.startsBefore(otherBlockSlice) ? otherBlockSlice : blockSlice;
+
+    // An illegal move is when a predicate in a later block depends on an earlier block.
+    // Get the questions used in any predicate in the latter block.
+    ImmutableSet<Long> predicateQuestions =
```

Comment:
(non-blocking) would it be helpful if we have a method that checks for valid predicate ordering on a program e.g. `ProgramDefinition#hasValidPredicateOrdering()`, and then check after each move if it is still valid. I'm assuming that method would be helpful in a few other places. But if not, then forget about it!

If you did this, the check would have to move right before the return statement.

Reply:
I like your idea! Added `hasValidPredicateOrdering` and used it in `moveBlock`